cadc-storage-adapter-fs/src/main/java/org/opencadc/inventory/storage/fs/FileSystemStorageAdapter.java

```diff
@@ -317,40 +352,65 @@ public StorageMetadata put(NewArtifact newArtifact, InputStream source)
         InventoryUtil.assertNotNull(FileSystemStorageAdapter.class, "artifact", newArtifact);
         InventoryUtil.assertNotNull(FileSystemStorageAdapter.class, "source", source);
         
-        Path path = null;
-        Path existing = null;
+        Path txnPath = null;
+        Path contentPath = null;
+        Path txnTarget = null;
         StorageLocation storageLocation = null;
         URI artifactURI = newArtifact.getArtifactURI();
+        log.debug("put: artifactURI: " + artifactURI.toString());
         
         try {
-            
+            // Make storage location using artifactURI
             storageLocation = this.createStorageLocation(artifactURI);
-            path = this.createStorageLocationPath(storageLocation);
-            
-            if (Files.exists(path)) {
+            StringBuilder path = new StringBuilder();
+            path.append(TXN_FOLDER + "/");
+            log.debug("resolving transaction path: " + path.toString());
+            txnPath = root.resolve(path.toString());
+
+            // add UUID to txnPath to make it unique
+            txnTarget = txnPath.getParent().resolve(UUID.randomUUID().toString());
+            log.debug("resolved txnTarget file: " + txnTarget);
+
+            if (Files.exists(txnTarget)) {
+                // This is an error as the name in the transaction directory should be unique
+                log.debug("file/directory exists");
+                throw new IllegalArgumentException(txnPath + " already exists.");
+            }
+
+            Files.createDirectories(txnTarget.getParent());
```

Comment:
Should never need to createDirectories for txnTarget

Reply:
ok